programs/perpetuals/src/state/stake.rs

```diff
@@ -0,0 +1,173 @@
+//! Stake state and routines
+//!
+//! Stake represent the LM staking account of a user of the platform.
+//! Staking of LM token grant access to a share of the platform revenues
+//! proportionnal to the amount of staked tokens.
+//! To ensure fair distribution, rewards are per rounds.
+//! A round has a fixed minimum duration, after which it will be available for resolution.
+//! Resolution of a round closes it, define the amount of reward per staked token during that round,
+//! and initialize the next staking round.
+//!
+//! User can claim their `Stake`, by doing so the program will read the vec of `StakeRound`s in the `Cortex`
+//! and determined based on the `Stake.inception_timestamp` if the user is elegible for the round rewards.
+//! The `StakeRound` will increase it's `token_claim` property, and once it matches the `token_stake` one,
+//! will remove itself from the record.
+//!
+//! Since there is a hard limitation on the data stored onchain on solana (10mb per accounts), the `stake_rounds`
+//! property of the `Cortex` have a upper limit. Once the limit is nearing, the `claim_stake` for `Stake`
+//! where the `inception_timestamp` is old enough will offer % of the reward to the caller, similar to a liquidation.
+//!
+//! This should ensure that the `stake_rounds` vec never grow beyond what's storable, in a decentralized fashion.
+//! (Adrena will run a claim-bot until decentralized enough, but anyone can partake)
+//!
+
+use {
+    super::{
+        cortex::{StakingRound, DAYS_PER_YEAR, HOURS_PER_DAY, SECONDS_PER_HOURS},
+        perpetuals::Perpetuals,
+    },
+    crate::math,
+    anchor_lang::prelude::*,
+    std::cmp::max,
+};
+
+#[account]
+#[derive(Default, Debug)]
+pub struct Stake {
+    pub amount: u64,
+    pub bump: u8,
+    // this value is refreshed during each call to add_stake, remove_stake or claim_stake
+    pub stake_time: i64,
+}
+
+// define if the additionnal caller on claim_reward is eligible for bounty
+pub enum BountyStage {
+    NoReward,
+    StageOne,
+    StageTwo,
+}
+
+impl BountyStage {
+    const fn reward_bps(self) -> u64 {
+        match self {
+            BountyStage::NoReward => 0,
```

Comment:
We created this feature to users not lose theirs rewards due to solana technical limitations.

I was thinking we could use it also for an other purpose. We could add an extra Stage to incentive claimor to claim all the times, effects would be:
- Users receives rewards every 6h (streamflow like)
- Onchain account stay small

I imagine users may not want to lose rewards, they could set a var when staking "auto-claim" which reward claimor or not.

Reply:
True, interesting feature! I don't think we should bother with this for now, especially given the tokenomics being in the work, but def something I'd like to see, can also mix in some auto compound